### ISSUE TITLE
Updates Requested

### ISSUE THREAD

**Issue by JPRuskin** [TomasHubelbauer/ps-remoting#1]
I was looking for ways to remote into a Windows Sandbox VM, and found this.

I tried `Enable-PSRemoting` within the sandbox, and got the following error:

![image](https://user-images.githubusercontent.com/1975761/230391968-1c39648e-7eaf-44b7-847a-8a6f786729a3.png)

I then ran `Get-NetConnectionProfile | Set-NetConnectionProfile -NetworkCategory Private`, though there may be a configuration possible in the Sandbox definition to address this, and then `Enable-PSRemoting -Force`. You could also use `-SkipNetworkProfileCheck`, of course.

After this, running `Invoke-Command` (or `Enter-PSSession`) against the machine worked (on my box, at least).

![image](https://user-images.githubusercontent.com/1975761/230392854-6f3edc98-156d-4858-af00-a4c2849e3cad.png)

I suspect that you may need to add something to your trusted hosts, such that you can authenticate. Have a look at [something like this](https://www.dtonias.com/add-computers-trustedhosts-list-powershell/).

The credential you should use is provided to the WindowsSandboxClient as process arguments (and I haven't yet figured out how to run Sandbox headlessly) - you can retrieve it by viewing the process in procexp.exe (or task manager), or with PowerShell, as below:

```
if ((Get-Process WindowsSandboxClient).CommandLine -match "<AccountUser>(?<User>.+)</AccountUser><AccountPassword>(?<Password>.+?)</") {
$Credential = [PSCredential]::new(
    $Matches.User,
    ($Matches.Password | ConvertTo-SecureString -AsPlainText -Force)
)
} else {Write-Error "Didn't find a match"}
```

Unless folk change the disk image being used for the sandbox, it's unlikely to ever use a different username, but may as well grab it out!

The only problem remaining, then, is getting the machine name or connection address out. Connecting to the container ID specified in the arguments passed to WindowsSandboxClient doesn't work (but would have solved a lot of problems). I suspect it's not being treated like that sort of container.
We could mount a directory, and have the init commands for the Sandbox write a file to that location, but there should be a better way!

Sorry if you'd already solved all of this! I'd be interested to hear if you're already past all of this, or if there were any updates for this repository.

**Comment by arc-hbas** [comment]
Looks like the that `WindowsSandboxClient` process is not created directly any longer. The only three processes created as of today are:

> vmmemWindowsSandbox
> WindowsSandboxRemoteSession.exe
> WindowsSandboxServer.exe

None of which leak anything significant into cmd line args.

So looks like right now our two options are to either:

<ul>
     a) create/modify existing user with our own password (see <a href="https://superuser.com/a/1697484/1729376">this SO</a> which re-enables administrator account and allows blank password usage), or
</ul>
<ul>
     b) change password for the default WDAGUtilityAccount user. 
</ul>

Note while both work, b) could lead to some unintended side-effects so a) should be preferred. In particular with a), what I noticed is that if you need to reboot the system (which are persisted now) or toggle any of the settings options (i.e. clipboard redirect, etc), the sandbox GUI will not be able to re-connect. This is a strong indication that the password is still stored somewhere on host, but not quite sure where exactly (memory?). Using a diff account should avoid this headache.

Side note: looks like there is this [wsb cli](https://learn.microsoft.com/en-us/windows/security/application-security/application-isolation/windows-sandbox/windows-sandbox-cli) tool now, new as of 24H2 (Oct '24). This gives us the ability to execute some basic commands against the sandbox (start, stop, etc). In particular interest are these three:

```
wsb list
wsb ip --id {ID}
wsb exec --id {ID} -c {CMD} -r {RUNAS}
```

First one allows us to get the currently running sandbox's ID. Second one allows us to take that ID and get it's reachable IP address. And the third one allows us to execute any command remotely on the sandbox directly from the host. Note that the `exec` command doesn't return any input (probably by design), so it just returns whether the command worked or not (based on exit code).

This would allow us to a) quickly retrieve the sandbox IP, b) exec a command to change a user's password, & c) use the IP and new password within our host script to start a PS remote session (assuming it's already enabled on sandbox).

Would still really like to know if anyone has any tricks for retrieving the current sandbox's password.

**P.S.** If using that `wsb` cli tool, one critical thing I noticed was that running the `wsb list` cmd in an elevated shell did NOT produce any results. If however you open a non-elevated shell, then you would see the corresponding sandbox ID. This got me curious if I could start a sandbox elevated (right-click run as admin), which does work. Even more curious, you can run both of these side-by-side, one elevated and one non-elevated, although what I found was that only the first one would show in the GUI, the second one would error with `The remote environment is Logging off. The connection is closed.` (if you swap the open order, it always has a problem with the second one). Consequently, that means closing the sandbox window with the error will stop and clear out that sandbox, making this find apparently useless. EXCEPT, you can start a sandbox from the cli using `wsb start` now w/o connecting to it! This will keep the sandbox running until you issue a `wsb stop --id {id}` command. If you try to connect to it with `wsb connect --id {id}`, you will still receive the same error, but this time the sandbox won't be stopped and remains persisted (check `wsb list`). Therefore, if you need an isolated windows environment to run some scripts w/o a GUI, you can enable PS remoting via `wsb exec` and be on your way. Granted you could probably enable RDP within the sandbox for a GUI experience if you really wanted to, but I have not confirmed if this would work yet.

I doubt this is an intended use case, but sure makes things handy to 1) persist sandbox containers longer than the UI interaction, and b) optionally run two sandbox containers simultaneously (depending on use case).

Sorry got a bit off topic there 😅...

**Comment by arc-hbas** [comment]
Ok it was bugging me not knowing where that password was coming from and how to pull it out. So I opened VS and attached to the sandbox GUI process and learned some things.

**TLDR; There is a GRPC endpoint setup by the `WindowsSandboxServer.exe` process (I believe). We can use this to retrieve the password for a given sandbox provided we have the correct sandbox id.**

We can get the configuration ourself using the below script:

```ps1
using assembly "{path_to}\MicrosoftWindows.WindowsSandbox_0.4.31.0_x64__cw5n1h2txyewy\SandboxCommon.dll"

$client = [SandboxCommon.Grpc.GrpcClient]::new()

$config = $client.GetRdpClientConfigAsync([guid]'{SANDBOX_GUID}').Result

$config.Password # just a guid
```

**_NOTE_**: By default, windows sandbox is installed as a windows app and is a present in a directory like so: `C:\Program Files\WindowsApps\MicrosoftWindows.WindowsSandbox_0.4.31.0_x64__cw5n1h2txyewy`. This is problematic because that folder is restricted and even an elevated shell can't access it. I'm sure we could modify the permissions of those folders to get access, but for testing it was just simpler to copy the files from that directory to our own and run the script based on that. As you can see, we're only loading the `SandboxCommon.dll` file, so I suspect we don't need 99% of the rest of the files in that folder.

Realistically, we shouldn't depend in that app folder at all. Potentially a safer solution would be to use [Grpc.Net.Client](https://www.nuget.org/packages/Grpc.Net.Client) directly and create our own client that emulates the same proto as the sandbox client. While looking over the code, I saw nothing sophisticated enough that couldn't be implemented in a separate app.

One tricky bit is the fact that the GRPC connection happens over a named pipe. You can see these by running the below PS command. Just make sure the sandbox is currently running (note that you will see two of them if you're running an elevated AND non-elevated sandbox simultaneously like I described in my last reply). If the sandbox is closed/destroyed, then the named pipe will also be closed. The path of the named pipe will look something like this:

```ps1
PS> (get-childitem \\.\pipe\*wsandbox*).FullName
\\.\pipe\wsandbox\{USER_SESSION_GUID}
```

The user session GUID can be derived like so:

```ps1
# private UserSessionId: [SandboxCommon.Grpc.GrpcServer]::UserSessionId
# public PipePath: [SandboxCommon.Grpc.GrpcServer]::PipePath
$owner = [System.Security.Principal.WindowsIdentity]::GetCurrent().Owner.ToString()
$UserSessionId = [guid]::new([System.Security.Cryptography.MD5]::HashData([System.Text.Encoding]::UTF8.GetBytes($owner))).ToString()
```

**_Note_:** The user session guid seems to be specific to the current user. It's derived from the owning SID (security identifier). You'll see that the SID for the same user differs depending on whether running from an elevated or non-elevated shell, which may explain why we're able to run two sandboxes simultaneously. Matter of fact, this probably means I can create an adhoc user and open a command prompt as them (with `runas`) and create another sandbox via `wsb` cli. I did try this using the built-in "NT AUTHORITY\SYSTEM" account (via `PsExec`) but ran into access errors. Anyhow, problem for another day...

**_Note_**: I noticed that if you do initiate a reboot, the sandbox will have both a new IP assigned to it as well as a new password. The above `GetRdpClientConfigAsync` command does reflect the new password pretty quickly during the reboot process though (although the IP is still blank for whatever reason, so just use `wsb ip` for that).

...

Sorry for hijacking the original issue in regards to powershell remoting, although all of this is somewhat related for a full dynamic solution. Bottom line is now we can enable PS remoting from host, find the auto-created password & ip, and invoke remote command(s). Putting it all together:

```ps1
using assembly "{path_to}\MicrosoftWindows.WindowsSandbox_0.4.31.0_x64__cw5n1h2txyewy\SandboxCommon.dll"

wsb start

[guid]$sandboxId = wsb list

$client = [SandboxCommon.Grpc.GrpcClient]::new()

$config = $client.GetRdpClientConfigAsync($sandboxId).Result

wsb exec --id $sandboxId -r SYSTEM -c 'powershell -c Enable-PSRemoting -force -SkipNetworkProfileCheck'

$sandboxIp = wsb ip --id $sandboxId

$cred = New-Object System.Management.Automation.PSCredential ("WDAGUtilityAccount", (ConvertTo-SecureString ($config.Password) -AsPlainText -Force))

Invoke-Command -ComputerName $sandboxIp -Credential $cred -ScriptBlock { Write-Host "Hello from sandbox!"; whoami; hostname; }

wsb stop --id $sandboxId
```

Hope this helps someone else! 🍻

**Comment by arc-hbas** [comment]
My theory on being able to run another sandbox under a different user via CLI is partially correct. 

**_TLDR: got it to work w/ `NT AUTHORITY\SYSTEM` user (via `PsExec`), but not any others._** 

I tried creating a local user, but it couldn't access the sandbox app folder by default, whether I put the user in a admin group or not. I also tried the other built-in `LocalService`/`NetworkService` accounts, but both also had access denied issues (prob for the best). I'm sure some simple tricks could be done to make it happen (like copying the exe/dll's elesewhere), but regardless looks like we have the ability to spin-up three total sandboxes under the current user as-is, two of which would need to be headless (w/o enabling RDP at least). Should work for most cases I'd bet.

![Image](https://github.com/user-attachments/assets/8d583162-e776-45a7-8618-b37e53348e57)

![Image](https://github.com/user-attachments/assets/0ee93b53-b3cd-415f-a2b7-0ea030f9c0df)

**Comment by TomasHubelbauer** [comment]
Thanks! This is awesome research. I've switched to macOS since creating this repository so somewhat awkwardly I really have nothing to add here, I'm sorry, but I'll be happy to merge whatever solution this research becomes a stepping stone for.

**Comment by gerneio** [comment]
Thanks, appreciate it @TomasHubelbauer! Wonder if [parallels](https://www.parallels.com/pd/windows-on-mac/) would allow windows sandbox's on ur mac 🤔.

FYI, finally created my own project breaking down all my findings, as well as the new ones:

https://github.com/gerneio/WindowsSandboxPlayground

**Comment by TomasHubelbauer** [comment]
As someone who hates Docker I really resonated with the readme of your project 😂 Thanks for sharing your knowledge <3
